tool/src/dart/mod.rs

```diff
@@ -330,6 +334,100 @@ impl<'a, 'cx> TyGenContext<'a, 'cx> {
+        // stuff makes this complicated
+
+        // Lifetimes actually used in the output
+        let used_method_lifetimes = method.output.used_method_lifetimes();
+        let mut method_lifetimes_map: BTreeMap<_, _> = used_method_lifetimes
+            .iter()
+            .map(|lt| {
+                (
+                    *lt,
+                    MethodLifetimeInfo {
+                        incoming_edges: Vec::new(),
+                        all_longer_lifetimes: method
+                            .lifetime_env
+                            .all_longer_lifetimes(lt)
+                            .collect(),
+                    },
+                )
+            })
+            .collect();
+
+        // Add a parameter to the method_lifetimes map for any lifetimes it references.
+        //
+        // This basically boils down to: For each lifetime that is actually relevant to borrowing in this method, check if that
+        // lifetime or lifetimes longer than it are used by this parameter. In other words, check if
+        // it is possible for data in the return type with this lifetime to have been borrowed from this parameter.
+        // If so, add code that will yield the ownership-relevant parts of this object to incoming_edges for that lifetime.
+        let mut add_param_to_map = |ty: &hir::Type, param_name: &str| {
+            if used_method_lifetimes.is_empty() {
+                return;
+            }
+
+            // Structs have special handling: structs are purely Dart-side, so if you borrow
+            // from a struct, you really are borrowing from the internal fields.
+            if let hir::Type::Struct(s) = ty {
+                let def = s.resolve(self.tcx);
+                for method_lifetime in method_lifetimes_map.values_mut() {
+                    // Note that ty.lifetimes()/s.lifetimes() is lifetimes
+                    // in the *use* context, i.e. lifetimes on the Type that reference the
+                    // indices of the method's lifetime arrays. Their *order* references
+                    // the indices of the underlying struct def. We need to link the two,
+                    // since the _fields_for_lifetime_foo() methods are named after
+                    // the *def* context lifetime.
+                    //
+                    // Concretely, if we have struct `Foo<'a, 'b>` and our method
+                    // accepts `Foo<'x, 'y>`, we need to output _fields_for_lifetime_a()/b not x/y.
+                    let def_lifetimes = def.lifetimes.all_lifetimes();
+                    let use_lifetimes = s.lifetimes.lifetimes();
+                    assert_eq!(def_lifetimes.len(), use_lifetimes.len(), "lifetimes array found on struct def must match lifetime parameters accepted by struct");
+                    // the type lifetimes array
+                    for (def_lt, use_lt) in def_lifetimes.zip(use_lifetimes) {
```

Comment:
thought: this trick works for structs but not for borrowing opaques, which can have a `self` lifetime in `use_lifetimes`